framework/src/engine/legacy/legacy_chain_handler.ts

```
@@ -12,56 +12,202 @@
+	 * Make a request to that random peer by calling its `getLegacyBlocksFromId` method with `data` property set to `legacyBlock.header.id`
+	 * Try to decode response data buffer into Block, apply penalty in case of error
+	 * Start saving parsed blocks one by one
+	 * repeat, if last parsed block height is still higher than bracket.startHeight
+	 * finally, update node info
+	 */
+	// eslint-disable-next-line @typescript-eslint/member-ordering
+	public async syncBlocks(bracket: LegacyBlockBracket, legacyBlock: LegacyBlock): Promise<void> {
+		const connectedPeers = (this._network.getConnectedPeers() as unknown) as Peer[];
+		const peersWithLegacyInfo = connectedPeers.filter(
+			peer =>
+				!!(peer.options as { legacy: Buffer[] }).legacy.find(snapshotBlockID =>
+					snapshotBlockID.equals(Buffer.from(bracket.snapshotBlockID, 'hex')),
+				),
+		);
+		if (!peersWithLegacyInfo) {
+			throw new PeerNotFoundWithLegacyInfo('No peer found with legacy info.');
+		}
+
+		const randomPeerIndex = Math.trunc(Math.random() * peersWithLegacyInfo.length - 1);
+		const { peerId } = peersWithLegacyInfo[randomPeerIndex];
+
+		const p2PRequestPacket: P2PRequestPacket = {
+			procedure: 'getLegacyBlocksFromId',
+			data: legacyBlock.header.id,
+		};
+
+		const response = await this._network.requestFromPeer({ ...p2PRequestPacket, peerId });
+		if (!(response.data as []).length) {
+			return;
+		}
+
+		// `data` is expected to hold blocks in DESC order
+		const { data } = response;
+		let legacyBlocks: LegacyBlock[];
+		try {
+			// this part is needed to make sure `data` returns ONLY `{ blocks: Buffer[] }` & not any extra field(s)
+			const { blocks } = codec.decode<{ blocks: Buffer[] }>(
+				getBlocksFromIdResponseSchema,
+				data as Buffer,
+			);
+			this._applyValidation(blocks);
+			legacyBlocks = blocks.map(block => decodeBlock(block).block);
+		} catch (err) {
+			this._network.applyPenaltyOnPeer({ peerId, penalty: 100 });
+			return;
```

I think here it should throw error instead of return otherwise , it will continue to the syncing next bracket without finishing this bracket

---

so in general this code block should be like,
prefer this being called and applied penalty separately,

- check blocks.length > 0, apply penalty, log that penalty was applied to a peer and call sync() again
- if the above check passes, call _applyValidation, log that penalty was applied to a peer, if fails then apply penalty and call sync() again